shared_model/utils/variant_deserializer.hpp

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_VARIANT_DESERIALIZER_HPP
+#define IROHA_VARIANT_DESERIALIZER_HPP
+
+#include <boost/serialization/variant.hpp>
+
+namespace shared_model {
+  namespace detail {
+    template <class S>
+    struct variant_impl {
+      struct load_null {
+        template <class Archive, class V>
+        static void invoke(Archive &, int, V &) {}
+      };
+
+      struct load_impl {
+        template <class Archive, class V>
+        static void invoke(Archive &ar, int which, V &v) {
+          if (which == 0) {
+            using head_type = typename boost::mpl::front<S>::type;
+            v = head_type(ar);
```

Comment:
This is means that the variant itself should have a constructor that deserialize from the proto.
I think this is a bad way, and not really flexible one. It's possible to delegate that work to special class (let say deserializer) to perform that work with the same compile-times checks:
```cpp
using head_type = typename boost::mpl::front<S>::type;
v = deserializer<head_type>::load(ar);
```

and somewhere:
```cpp
template<>
class deserializer<SomeType> {
  static SomeType load(Proto p) {
    return SomeType{p.a(), p.b(), ...};
  }
}
```

[Here's the code](https://gist.github.com/l4l/5b8d7ee63fa1dbd6535faadbc5f37991) to play with

Reply:
If we will have different behavior on deserialization, we will use partial specialization instead. In current PR it will lead to a lot of redundant code.